NTS support for systemd-timesyncd

squell opened the pull request:
This is intended to eventually close #9481

Some comments:

* To keep the modification well-separated, the additions to timesyncd are split up two clearly distinguishable parts:
  - The NTS logic itself; these were developed first in the repo https://github.com/squell/nts
  - Modifications to integrate those into the `timesyncd` sources. 

* Currently, to select an "NTS server" an `NTS=` line needs to be added to the configuration; because the intent is "secure time", enabling it will ignore all other NTP servers ~~(except for the `FallbackNTP`)~~. Maybe in the future (if NTS gains traction) it might be wise to configure this differently (e.g. `RequireSureTime=yes` or something).

* ~~The function `manager_nts_obtain_agreement` currently does busy-waiting (which has the advantage of making it perhaps a bit easier to follow the logic). This still **should be integrated** in the async event loop! (hence the "draft" status of this PR).~~

* ~~The current meson build system modifications take some pains to make the selection of TLS and crypto backends flexible. I've learned from @poettering that the intent is to consolidate on OpenSSL; so that part can be simplified. Do note that the minimum required OpenSSL version is v3.5 due to a bug in OpenSSL (and a workaround for that bug not being universally compatible with existing NTS servers). In any case, this PR still allows selecting `gnutls` for TLS and `nettle` and `gcrypt` as alternative crypto backends. Not included in this PR (but available if desired) is a crypto backend that relies on [libaes_siv](https://github.com/dfoxfranke/libaes_siv), which would allow using older OpenSSL versions.~~

* I'm not 100% sure I didn't miss some config file for integrating the tests and fuzzing code, if there's anything missing, a PR on our forked repo or *Suggestion* is appreciated!

* To get in the weeds on NTS protocol details:
   - This patch set supports plenty of AEAD choices. This PR will signal a preference for `AES-GCM-SIV` crypto (because it is most space-efficient). This will only work with up-to-date Chrony implementations (4.6.1 or later, [for annoying reasons](https://chrony-project.org/doc/spec/nts-compliant-128gcm.html)). In my practical tests I haven't found any NTS servers running an older Chrony version, and I don't think it's very useful to make this fresh NTS implementation "bug-compatible" with older Chrony versions. By the time NTS lands in `timesyncd` I would hope that this issue is long forgotten. But if this is seen as problematic, it's very easy to make the default `AES-SIV-CMAC`.
   - In case there are missing cookies, the code will only ever replenish these "one cookie at a time". I've discovered (and been advised by @davidv1992) that many routers don't like large NTP packets, so this is the more robust option. 
   - In NTS it's possible to request additional cookies by adding a placeholder field in the encrypted part of the NTP request and/or in the unencrypted part; right now this code will only do it in the unencrypted part. Some practical testing is required but I believe not all NTS servers out there like it when the "authenticated and encrypted extension field" contains a "Cookie Placeholder" (or in fact anything else) and the current set-up is the most compatible choice.

squell committed:
let the user configure NTS servers using NTS= / SecureServers=

squell committed:
let the user configure NTS servers using NTS= / SecureServers=

squell committed:
add logic to try NTSKE servers first (and only)

squell committed:
add logic to try NTSKE servers first (and only)

squell committed:
add initial key exchange control flow logic

squell committed:
add initial key exchange control flow logic

squell committed:
add error handling

squell committed:
add error handling

squell committed:
store keys and cookies

squell committed:
store keys and cookies

squell committed:
send extension fields

squell committed:
send extension fields

squell committed:
add tolerance for NTS servers that send broken-up NTS responses

squell committed:
add tolerance for NTS servers that send broken-up NTS responses

squell committed:
don't add nonce if there are NTS extension fields

squell committed:
don't add nonce if there are NTS extension fields

squell committed:
free cookies, move handshake out of manager object, fix some compilation warnings

squell committed:
free cookies, move handshake out of manager object, fix some compilation warnings

squell committed:
add verification of NTS extension fields

squell committed:
add verification of NTS extension fields

squell committed:
add detection of NTS unique identifier

squell committed:
add detection of NTS unique identifier

squell committed:
add FIFO queue system for cookies

squell committed:
add FIFO queue system for cookies

squell committed:
request additional cookies if we are missing some

squell committed:
request additional cookies if we are missing some

squell committed:
change classification of log messages

squell committed:
change classification of log messages

squell committed:
add cookie choice randomization

squell committed:
add cookie choice randomization

squell committed:
add timeout logic

squell committed:
add timeout logic

squell committed:
scaffolding for meson

squell committed:
scaffolding for meson

squell committed:
make timeout configurable

squell committed:
make timeout configurable

squell committed:
add a preference for GCM

squell committed:
add a preference for GCM

squell committed:
add comment about cookie replenishment strategy

squell committed:
add comment about cookie replenishment strategy

squell committed:
integrate nts config into existing systemd structure (mostly)

squell committed:
integrate nts config into existing systemd structure (mostly)

squell committed:
add NTS-tests

squell committed:
add NTS-tests

squell committed:
add fuzzers

squell committed:
add fuzzers

squell committed:
add sources from https://github.com/squell/nts.git

squell committed:
add sources from https://github.com/squell/nts.git

squell committed:
set socket to O_NONBLOCK

squell committed:
set socket to O_NONBLOCK

squell committed:
window dressing: explicitly mention security used in the log

squell committed:
window dressing: explicitly mention security used in the log

squell committed:
add manpage for NTS

squell committed:
add manpage for NTS

squell committed:
mark NTSKeyExchangeServers the same as FallbackNTPServers

squell committed:
mark NTSKeyExchangeServers the same as FallbackNTPServers

squell committed:
remove stray line

squell committed:
remove stray line

squell committed:
silence 'this is not a string' warning in fedora builds

squell committed:
silence 'this is not a string' warning in fedora builds

squell committed:
silence invalid warnings on some other platforms

squell committed:
silence invalid warnings on some other platforms

poettering commented:
hmm, what's the license of the code you are copying in in the first commit? we really want spdx headers.

poettering commented on meson.build:
```diff
@@ -1679,6 +1679,35 @@ conf.set10('ENABLE_SSH_USERDB_CONFIG', conf.get('ENABLE_USERDB') == 1 and sshdco
+                have_gnutls = false
+        else
+                have_gnutls = conf.get('HAVE_GNUTLS') == 1 and libgnutls.version().version_compare('>= 3.6.0')
+        endif
```
please no new code with gnutls. just use openssl

squell replied:
I guess that also holds for gcrypt? That would simply the build config greatly.

DaanDeMeyer replied:
Yes just openssl please

poettering commented on src/timesync/fuzz-nts-extfields.c:
```diff
@@ -0,0 +1,80 @@
+#include <stdio.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <unistd.h>
```
please sort includes, as per coding style 

see https://systemd.io/CODING_STYLE/

poettering commented on src/timesync/fuzz-nts-extfields.c:
```diff
@@ -0,0 +1,80 @@
+#include "nts_extfields.h"
+
+static void eat(volatile const uint8_t* buf, size_t size) {
+        if (buf) while (size) (void)buf[size--];
```
this is not our coding style. please never stuff so much stuff in one line. this needs two line breaks in the middle.

also, this loop will be optimized away by any compiler, i don't get it?

poettering commented on src/timesync/fuzz-nts-extfields.c:
```diff
@@ -0,0 +1,80 @@
+        int len = size > sizeof(buffer)? sizeof(buffer) : size;
+
+        memcpy(buffer, data, len);
+        if (len < 48) return 0;
```
coding style, add line break

poettering commented on src/timesync/fuzz-nts-extfields.c:
```diff
@@ -0,0 +1,80 @@
+        memcpy(buffer, data, len);
+        if (len < 48) return 0;
+
+        struct NTS_Query nts = (struct NTS_Query) {
```
the `(struct NTS_Query)` is redundant

poettering commented on src/timesync/fuzz-nts-extfields.c:
```diff
@@ -0,0 +1,80 @@
+        };
+
+        /* fuzz the NTS extension field parser */
+        struct NTS_Receipt rcpt = { 0, };
```
please just use `= {};`, i.e. empty brackets. That is only allowed as a gnu extension, but it is one we use heavily.

poettering commented on src/timesync/fuzz-nts-extfields.c:
```diff
@@ -0,0 +1,80 @@
+
+#define BLKSIZ 16
+
+int NTS_encrypt(uint8_t *ctxt,
```
the code below assumes that ctxt is 16 byte long hence write this as `uint8_t ctxt[static BLKSIZ]`

poettering commented on src/timesync/nts_crypto_nettle.c:
```diff
@@ -0,0 +1,200 @@
+};
+
+const struct NTS_AEADParam* NTS_get_param(NTS_AEADAlgorithmType id) {
+        for (size_t i=0; i < ELEMENTSOF(supported_algos); i++)
```
FOREACH_ELEMENT()

poettering commented on src/timesync/nts_tls.c:
```diff
@@ -0,0 +1,247 @@
+                return -2;
+
+        for (int i=0; i < 2; i++) {
+                const char context[5] = { 0, 0, (aead >> 8) & 0xFF, aead & 0xFF, i };
```
this is really weird. you use "char" here, which is signed on most systems. but you then assign some value & 0xFF to it, which suggests you mean the 0…255 range. thus you are always triggering an overflow here. use uint8_t hence

poettering commented on src/timesync/tcp_connect.c:
```diff
@@ -0,0 +1,59 @@
+
+                int flags;
+                if((flags = fcntl(sock, F_GETFL)) < 0 || fcntl(sock, F_SETFL, flags | O_NONBLOCK) < 0) {
+                        return close(sock), -2;
```
ever use raw close() in our codebase. use safe_close(). 

and usually you don't want to use that either, use `_cleanup_close_`. 

finally, please just create the socket with O_NONBLOCK right-away, u.e. use SOCK_NONBLOCK.

poettering commented on src/timesync/tcp_connect.c:
```diff
@@ -0,0 +1,59 @@
+                }
+
+                int status = 1;
+                setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, &status, sizeof(status));
```
please debug log about unexpected failures at least. also, use setsockopt_bool()

poettering commented on src/timesync/tcp_connect.c:
```diff
@@ -0,0 +1,59 @@
+        hints.ai_socktype = type;
+
+        struct addrinfo *info;
+        if (getaddrinfo(host, NULL, &hints, &info) != 0)
```
please never use a blocking call like this in an event-loop based serice such as timesyncd. use sd_resolve_xyz

poettering commented on src/timesync/timesyncd-manager.c:
```diff
@@ -1290,3 +1449,154 @@ int bus_manager_emit_ntp_server_changed(Manager *m) {
+        log_debug("Performing key exchange with %s\n", m->current_server_name->string);
+
+        NTS_TLS *nts_handshake = NTS_TLS_setup(m->current_server_name->string, socket);
+        assert_return(nts_handshake, -ENOMEM);
```
assert_return() should only be used to validate parameters of functions, and only if the caller is potentially not foreign, i.e. from some other codebase. or in other words: outside of libsystemd's public apis never use assert_return().

poettering commented on src/timesync/timesyncd-manager.c:
```diff
@@ -1290,3 +1449,154 @@ int bus_manager_emit_ntp_server_changed(Manager *m) {
+
+#if ENABLE_TIMESYNC_NTS
+
+static int manager_nts_obtain_agreement(Manager *m) {
```
we generally want that params are checked for basic validity, i.e. add `assert(m)` here

poettering commented:
i did some initial review now. please have a closer look at the coding style docs.

squell committed:
add more aggressive input validation

squell committed:
add more aggressive input validation

squell committed:
add spdx headers

squell committed:
add spdx headers

squell committed:
link source code to issues on forked repo

squell committed:
link source code to issues on forked repo

squell committed:
make nts logic depend on systemd headers

squell committed:
make nts logic depend on systemd headers

squell committed:
improve style in fuzz test, and use empty initializers gnuism

squell committed:
improve style in fuzz test, and use empty initializers gnuism

squell committed:
change interface of add_extension_fields to raise less eyebrows but still trigger the compiler in case of misuse

squell committed:
change interface of add_extension_fields to raise less eyebrows but still trigger the compiler in case of misuse

squell committed:
fix type of context

squell committed:
fix type of context

squell committed:
make enum more conforming to coding standard

squell committed:
make enum more conforming to coding standard

squell committed:
respect the include ordering guideline

squell committed:
respect the include ordering guideline

squell committed:
remove incorrect assert_return

squell committed:
remove incorrect assert_return

squell committed:
remove every crypto provider except openssl

squell committed:
remove every crypto provider except openssl

squell committed:
DRY null cipher

squell committed:
DRY null cipher

squell committed:
more safety around potential future OpenSSL misconfiguration

squell committed:
more safety around potential future OpenSSL misconfiguration

squell commented:
For my own reference, I'm tracking more general issues that still need attention on the forked repo: https://github.com/pendulum-project/nts-timesyncd/issues

squell committed:
silence more compiler warning

squell committed:
silence more compiler warning

bluca commented on src/timesync/meson.build:
```diff
@@ -29,31 +38,66 @@ else
+
+if conf.get('ENABLE_TIMESYNC_NTS') == 1
+
+executables += [
```
indentation and whitespace above

squell committed:
use a feature instead of manual combo

squell committed:
use a feature instead of manual combo

squell committed:
rewrite obtain_agreement to use the event manager

squell committed:
rewrite obtain_agreement to use the event manager

squell committed:
minor rewrite of index-into-buffer logic

squell committed:
minor rewrite of index-into-buffer logic

squell committed:
remove temporary global vars

squell committed:
remove temporary global vars

squell committed:
remove tcp_connect.c and replace with systemd's own socket handling

squell committed:
remove tcp_connect.c and replace with systemd's own socket handling

squell committed:
add key exchange timeout

squell committed:
add key exchange timeout

squell committed:
fix typos in documentation

squell committed:
fix typos in documentation

squell committed:
add version information for NTSKeyExchangeServers

squell committed:
add version information for NTSKeyExchangeServers

squell committed:
change error codes to standard ones

squell committed:
change error codes to standard ones

squell committed:
fix whitespace

squell committed:
fix whitespace

squell committed:
add fallthrough statements

squell committed:
add fallthrough statements

squell committed:
remove outdated comment, split expression

squell committed:
remove outdated comment, split expression

squell committed:
fixes in logic around socket opening/closing

squell committed:
fixes in logic around socket opening/closing

squell committed:
fix logic for retrying a server after NTS handshake succeeded

squell committed:
fix logic for retrying a server after NTS handshake succeeded

squell committed:
disable FallbackNTP when NTS= is configured

squell committed:
disable FallbackNTP when NTS= is configured

squell committed:
make socket connection non-blocking

squell committed:
make socket connection non-blocking

squell committed:
fix build with NTS disabled

squell committed:
fix build with NTS disabled

squell committed:
NTS servers are their own 'fallback' -- keep retrying them

squell committed:
NTS servers are their own 'fallback' -- keep retrying them

squell committed:
bump version number

squell committed:
bump version number

squell committed:
add comment

squell committed:
add comment

squell committed:
remove unused header includes

squell committed:
remove unused header includes

bluca commented on src/timesync/nts_crypto_openssl.c:
```diff
@@ -0,0 +1,194 @@
+typedef int init_f(EVP_CIPHER_CTX*, const EVP_CIPHER*, ENGINE*, const uint8_t*, const uint8_t*);
+typedef int upd_f(EVP_CIPHER_CTX*, uint8_t*, int*, const uint8_t*, int);
+
+static int process_assoc_data(
```
indentation all wrong below here

bluca commented on src/timesync/nts.h:
```diff
@@ -0,0 +1,141 @@
+
+#include "memory-util.h"
+
+#include <stddef.h>
```
use the "forward" headers instead when including in headers

squell committed:
style changes

squell committed:
style changes

squell committed:
remove 'ok()?' pattern from nts_tls

squell committed:
remove 'ok()?' pattern from nts_tls

squell committed:
remove check pattern from nts_packet

squell committed:
remove check pattern from nts_packet

squell committed:
remove check pattern from nts_extfields

squell committed:
remove check pattern from nts_extfields

squell committed:
remove check pattern from nts_crypto_openssl

squell committed:
remove check pattern from nts_crypto_openssl

squell committed:
replace ssl.h include with more direct headers (appease clang-tidy)

squell committed:
replace ssl.h include with more direct headers (appease clang-tidy)

squell committed:
track changes in nts code

squell committed:
track changes in nts code

squell committed:
first integration test (with external server)

squell committed:
first integration test (with external server)

squell committed:
add mock NTS server for integration tests

squell committed:
add mock NTS server for integration tests

squell committed:
make cookie management robust to bad NTS servers

squell committed:
make cookie management robust to bad NTS servers

squell committed:
integrate NULL'ing of TLS handle in close function

squell committed:
integrate NULL'ing of TLS handle in close function

squell committed:
don't wait for timeouts if we severed the connection early

squell committed:
don't wait for timeouts if we severed the connection early

squell committed:
add dbus property for inquiring NTS status

squell committed:
add dbus property for inquiring NTS status

Commit:
ignore SIGPIPE (can occur in local tests)

Commit:
ignore SIGPIPE (can occur in local tests)

Commit:
use systemd's headers and tabsize, add error-triggering options in mock-server

Commit:
use systemd's headers and tabsize, add error-triggering options in mock-server

Commit:
add error test cases

Commit:
add error test cases

squell committed:
skip integration tests if NTS not available

squell committed:
skip integration tests if NTS not available

squell committed:
run update-dbus-docs

squell committed:
run update-dbus-docs

squell committed:
new lints

squell committed:
new lints

squell committed:
make mock server show output more clearly

squell committed:
make mock server show output more clearly

squell committed:
papercut (makes connection error checking stricter)

squell committed:
papercut (makes connection error checking stricter)

squell committed:
add hack to integration test needed on Arch

squell committed:
add hack to integration test needed on Arch

squell commented on src/timesync/timesyncd.c:
```diff
@@ -157,6 +157,7 @@ static int run(int argc, char *argv[]) {
                 return r;
 
         umask(0022);
+        signal(SIGPIPE, SIG_IGN);
```
Otherwise one of the failure mode tests just makes systemd-timesyncd exit because of a broken pipe signal.

squell commented on src/timesync/timesyncd-manager.c:
```diff
@@ -460,11 +544,66 @@ static int manager_receive_response(sd_event_source *source, int fd, uint32_t re
+                        m->nts_missing_cookies--;
+                }
+
+                log_debug("NTP packet is authentic.");
```
Or of course fiddle with `NTS response` -> `Authenticated NTP response`. 💡

bluca replied:
mmh ok never mind then leave as is

squell committed:
incorporate feedback

squell committed:
incorporate feedback

squell committed:
suggestions in integration test

squell committed:
suggestions in integration test

squell committed:
invalidate all the cookie bytes (not all bits are significant for all servers)

also add defense in depth against replay (this shouldn't be possible anyway due to the logic
of the state machine)

squell committed:
invalidate all the cookie bytes (not all bits are significant for all servers)

also add defense in depth against replay (this shouldn't be possible anyway due to the logic
of the state machine)

squell commented on test/units/TEST-45-TIMEDATE.sh:
```diff
@@ -453,6 +479,173 @@ LOCAL"
+install_mock_certificate() {
+    servername="$1"
+    # this also installs a fake CA since we don't allow self-signed certificates
+    mkdir -p /etc/ssl/certs
```
Note: I've tried fiddling to get OpenSSL to accept a cert in /tmp by creating an override for SSL-related env vars but got stuck there. At least now the only thing left to cleanup is a good-for-nothing cert, and it gets removed -- other places also temporarily alter things in /etc in more dramatic ways. At least the test doesn't litter several things.

squell committed:
add authorship info to spdx comment

squell committed:
add authorship info to spdx comment

squell committed:
more claude feedback

squell committed:
more claude feedback

squell committed:
claude nits

squell committed:
claude nits

squell committed:
claude suggestions

squell committed:
claude suggestions

squell committed:
move NTS_TLS_close from manager_free to manager_disconnect

squell committed:
move NTS_TLS_close from manager_free to manager_disconnect

squell committed:
use correct fresh_cookie length in cookie management

squell committed:
use correct fresh_cookie length in cookie management

squell committed:
test was broken & other minor fixes

squell committed:
test was broken & other minor fixes

squell committed:
reverse claude suggestion; assert should happen first

squell committed:
reverse claude suggestion; assert should happen first

squell commented:
Uncaught fixes @claude didn't close in the second-to-last run:
* TLS session leak on EPOLLHUP/EPOLLERR: it does now unrefs in manager_disconnect
* SYNTHETIC_ERRNO(EINVAL) was already in place
* Summing return values risks signed overflow
* 127.0.0.01 not being a typo; this code screams "it's intentional" to me with a comment and an assert

Won't fixes:
* Assert before error check, this is intentional since failing the assert means OpenSSL did something terrible
* Cookie consumption; bad suggestion--NTS cookies aren't that precious.

This is ready for one more @claude review (which is catching some useful edge cases); I might do one more round of whack-a-mole but after that I'm taking a break of 1 or 2 weeks to avoid myopia and come back to this with a fresh mind.

squell committed:
claude suggestions

squell committed:
claude suggestions

poettering commented on src/timesync/nts.h:
```diff
@@ -0,0 +1,144 @@
+ *      0 upon success
+ *      negative upon failure (writes the error code to NTS_Agreement->error)
+ */
+int NTS_decode_response(uint8_t *buffer, size_t buf_size, struct NTS_Agreement *response);
```
the buffer should be const, no?

poettering commented on src/timesync/nts.h:
```diff
@@ -0,0 +1,144 @@
+ * RETURNS
+ *      A pointer to a ready-to-use TLS session, NULL upon failure (and then the error is stored in NTS_TLS_error)
+ */
+NTS_TLS* NTS_TLS_setup(const char *hostname, int socket);
```
i presume the 2nd arg is an fd? we usually call those parameters "fd" or "somethin_fd". hence maybe use "socket_fd" here?

poettering commented on src/timesync/nts.h:
```diff
@@ -0,0 +1,144 @@
+ *              -ENOBUFS not enough space in buffer
+ *              -EINVAL  unkown AEAD
+ */
+int NTS_TLS_extract_keys(NTS_TLS *session, NTS_AEADAlgorithmType aead, uint8_t *c2s, uint8_t *s2c, int key_capacity);
```
we count byte sizes *always* in size_t type, plesae do so here too.

i understand c2s and s2c are return parameters, right? then they should be called ret_c2s and ret_s2c

see coding style docs:

https://systemd.io/CODING_STYLE/#code-organization-and-semantics

poettering commented on src/timesync/nts.h:
```diff
@@ -0,0 +1,144 @@
+ * RETURNS
+ *      Nothing
+ */
+void NTS_TLS_close(NTS_TLS **session);
```
we typically call deep destructors "xyz_free()". please do so here too. And please define it not with a pointer-to-pointer argument, but with a simple pointer, and then make the function return NULL in the same type. i.e.:

```
ǸTS_TLS* NTS_TLS_free(NTS_TLS *session);
```

one would then use that like this:

```c
session = NTS_TLS_free(session);
```

poettering commented on src/timesync/nts_crypto.h:
```diff
@@ -0,0 +1,43 @@
+                const uint8_t *ptxt,
+                int ptxt_len,
+                const AssociatedData *info,
+                const struct NTS_AEADParam *aead,
```
drop the "struct"

poettering commented on src/timesync/nts_crypto.h:
```diff
@@ -0,0 +1,43 @@
+ * RETURNS: the number of bytes in the ciphertext (< 0 indicates an error)
+ */
+int NTS_encrypt(uint8_t *ctxt,
+                int ctxt_len,
```
(i mean, i am fully aware that openssl is sloppy there, and likes to use `int` as type for byte size values. in our codebase let's not propagate that sloppiness, and always use size_t when dealing in sizes or multiples of them

poettering commented on src/timesync/nts_extfields.c:
```diff
@@ -0,0 +1,278 @@
+        type = htobe16(type);
+        memcpy(buf->data, &type, 2);
+        len = htobe16(padded_len);
+        memcpy(buf->data+2, &len, 2);
```
use unaligned_write_be16() and similar helpers

poettering commented on src/timesync/nts_extfields.c:
```diff
@@ -0,0 +1,278 @@
+
+        return buf.data - dest;
+exit:
+        return 0;
```
please follow our usual error propagation here, i.e. return a nagetive errno on error. 

0 and positive return values are generally considered success in our tree, negative ones shall be errors.

poettering commented on src/timesync/nts_extfields.c:
```diff
@@ -0,0 +1,278 @@
+}
+
+/* caller checks memory bounds */
+static void decode_hdr(uint16_t *restrict a, uint16_t *restrict b, uint8_t *bytes) {
```
`const bytes[static 4]`?

squell committed:
misc comment

squell committed:
misc comment

squell committed:
lean on typedefs for everything

squell committed:
lean on typedefs for everything

squell committed:
NTS_TLS_close -> NTS_TLS_free

squell committed:
NTS_TLS_close -> NTS_TLS_free

squell committed:
change int -> size_t and add necessary bound checks and type casts

squell committed:
change int -> size_t and add necessary bound checks and type casts

squell committed:
use unaligned_read/write_be16 functions

squell committed:
use unaligned_read/write_be16 functions

squell committed:
replace 0 as error with negative error codes

squell committed:
replace 0 as error with negative error codes

squell committed:
introduce NTS_Identifier typedef

squell committed:
introduce NTS_Identifier typedef

squell committed:
remove not-quite-an-enum

squell committed:
remove not-quite-an-enum

squell committed:
add seperate definition header

squell committed:
add seperate definition header

squell committed:
re-use iovec for the things that are already iovecs

squell committed:
re-use iovec for the things that are already iovecs

squell committed:
replace slice with iovec in nts_extfields.c

squell committed:
replace slice with iovec in nts_extfields.c

squell committed:
use iovec instead of slice in nts_packet.c

squell committed:
use iovec instead of slice in nts_packet.c

squell committed:
bring test articles up to date

squell committed:
bring test articles up to date

squell committed:
appease coccinelle

squell committed:
appease coccinelle

squell commented:
I see that the integration test seems to be failing once more on Arch (whereas it succeeded a month ago). I can't tell from the logs what the problem is, but last time around it was that timesyncd simply refused to start because it couldn't tell whether a network was configured, which was fixed by 4545e9d316d80231843018dfc43c1d4e262d7462. My guess would be that it's something like that again.

But I can't test this locally since I seem to be unable to build a Arch mkosi image anymore, at some point running a "meson compile -C build mkosi" results in a pacman error:

> error: keyring is not writable

mkosi commit used: 23913f23a3c6cf2c883e3292a71a2ee723140389

I'd appreciate someone to take a look at this; there's a lot of variables that can cause mkosi build to fail, and of course the problem in the integration test might also require some knowledge of a different part of the whole systemd stack whereas I've been obviously been focussed on timesyncd--I don't think I'll efficiently figure this out
by burning a lot of time on this.

I did find one more nitfix that was hiding in my git stash that I'll push to give the CI one more go.

squell committed:
nitfix: add clamping to NTS_TLS_read/write

squell committed:
nitfix: add clamping to NTS_TLS_read/write

squell committed:
fix slice->iovec conversion in nts_extfields.c

i've also doubled checked the other slice->iovec conversions and this seems to be the only one where the base offset was missed

squell committed:
fix slice->iovec conversion in nts_extfields.c

i've also doubled checked the other slice->iovec conversions and this seems to be the only one where the base offset was missed

squell committed:
grammar

squell committed:
grammar

squell commented:
~~(Latest rebase was to absorb 8c3f64e61fd42541f56c5de1b0f926f72f9d935e)~~

(That didn't work, reverting to 6da8673a1c, TODO.)

squell committed:
track changes in 8c3f64e61fd42541f56c5de1b0f926f72f9d935e

squell committed:
track changes in 8c3f64e61fd42541f56c5de1b0f926f72f9d935e

squell committed:
EVP_ -> sym_EVP

squell committed:
EVP_ -> sym_EVP

squell committed:
SSL/BIO functions -> sym_

squell committed:
SSL/BIO functions -> sym_

squell committed:
claude: reorder starting connection and arming the timer

squell committed:
claude: reorder starting connection and arming the timer

squell committed:
claude nit: remove setsockopt_int from TLS socket

squell committed:
claude nit: remove setsockopt_int from TLS socket

squell committed:
add lifetime comment

squell committed:
add lifetime comment

squell committed:
remove copyright notices

(the parts that are completely new contributions are still copyright Trifecta Tech Foundation, but this makes future updates in the systemd branch easier)

user451421541757324 commented:
When will this be merged? It would be nice to have it ready in time for the next systemd release.

squell commented:
Note for the systemd maintainers: the time I have at Trifecta is getting very limited (it is part of an wider drive by Trifecta to promote NTS adoption). In recent times the effort to maintain the PR in a mergeable state with `main` (e.g. tracking regular updates to meson.build files, or wider changes such as the dynamic loading of OpenSSL) took more time than actually addressing feedback; so it would be good to know where the chances of this PR getting merged soon stand.

In any case, the state of affairs of this PR are:

- If the compile time support for NTS is turned off, or NTS is disabled in timesyncd.conf, I think these changes are very conservative and do not break existing timesyncd behaviour. I think that's the most important thing.

- The NTS additions have a bunch of unit- and fuzz-tests and I haven't found anything in those parts while working with that code outside of the context of systemd. I think the most tricky bits are in `timesyncd-manager.c`.

- Last time I checked I had the impression that the NTS integration test against the mock server regularly fails on ArchLinux (but not always). I think this could be related to timesyncd aborting because it can't detect a network interface and may not be related to this PR. For some time I've been unable to construct a `mkosi` vm Arch image so haven't been able to diagnose this intergration test. If this problem persists, *help would be appreciated*. However I don't think that this should be a blocker, since it's in a test that performs a full NTP query with a mock time server and such a test doesn't even exist without this PR; i.e. if a solution can't be found I would suggest temporarily disabling those tests on ArchLinux.

- I think the remaining Claude comment are a: an extreme nitpick for a local macro in test-only code, something that is not applicable, and something that has already been addressed.

On a personal note: I noticed that this PR was pointed at in [a recent discussion](https://github.com/systemd/systemd/issues/41085#issuecomment-4053124578) about using AI in systemd. From my perspective, while various parts of the Claude feedback have been very useful as a form of static analysis, in another sense Claude also exhibits some of the [review anti-patterns](https://www.chiark.greenend.org.uk/~sgtatham/quasiblog/code-review-antipatterns/) which has made it feel exhausting at times.

squell committed:
dlopen() openssl in unit tests

squell commented:
Note: for `test-nts-mockserver`, it also fails the new `link-abi` test. But:

- Making it pass would require adding rather a few defs to `shared/ssl-util.c` and `shared/crypto-util.h` which are only useful if one is making a TLS server (which I doubt are useful for any other part of systemd).
- At the same time, the first thing the mock server will need to do is dynamically load OpenSSL -- so why not let the operating system take care of that.

Making an exception for it seems like it's the more efficient solution rather than giving it the `sym_` + dlopen treatment as well. (And it's not part of the production code base--only exists to be used in an integration test.)

(CC'ing @daandemeyer since they are the author of d9600a2ac04)

poettering commented:
could you please squash the commits into logical steps rather than historical ones? it otherwise makes this very hard to review.

generally we expect PRs to consist of logical steps that can be bisected properly. ideally those are on the smaller side, but for bigger additions this is note often possible, so where necessary bigger commits are OK, but plesae split at least test stuff, and prep work in separate commits, and write proper commit msgs

poettering commented:
this currently does not pass CI, please rebase and make pass CI

poettering commented:
i did half a review now. but please make sure that CI passes, and make this reviewable by squashing commits, as suggested, otherwise it's very hrad for us to review.

squell commented:
> i did half a review now. but please make sure that CI passes, and make this reviewable by squashing commits, as suggested, otherwise it's very hrad for us to review.

To make this pass the CI I need a response to my earlier question about the nts-mock-server; it trips up the link-abi-test unit test (since it still links to OpenSSL) which causes the rest of the CI to not run. The easiest fix for that is to make an exception for nts-mock-server in that unit test, considering:

- It requires adding quite a bit of additional TLS symbols in the shared ssl-util/crypto-util headers that are only useful for writing a TLS server application, which I guess systemd itself will not be doing.
- It eats time on a chore that isn't achieving anything substantial, as the mock server is meant narrowly for integration tests.
- The very first thing the nts-mock-server would need to do is dlopen() OpenSSL anyway.

Also a repeated request: various comments in the last "half" review are very small things that are essentially highly local changes (e.g. whitespace usage, using `size_t` instead of `int` for small values). I would not mind simply receiving a commit containing the desired changes.

poettering commented:
i'll defer to @daandemeyer regarding the question whether dlopen() is necessary for that or not.

bluca commented:
The tests are packaged and shipped like other binaries, so yeah it needs the same dlopen treatment as all other components

squell commented:
Closing in favor of a future PR by @anthonyryan1, which will be a follow-up to this PR. See https://github.com/systemd/systemd/issues/9481#issuecomment-5057412501
